gemini/gradio-voice/app.py

```diff
@@ -0,0 +1,168 @@
+import asyncio
+import base64
+import pathlib
+from typing import AsyncGenerator, Literal
+
+from google import genai
+from google.genai.types import (
+    LiveConnectConfig,
+    PrebuiltVoiceConfig,
+    SpeechConfig,
+    VoiceConfig,
+)
+import gradio as gr
+from gradio_webrtc import AsyncStreamHandler, WebRTC, async_aggregate_bytes_to_16bit
+import numpy as np
+
+current_dir = pathlib.Path(__file__).parent
+
+
+def encode_audio(data: np.ndarray) -> str:
+    """Encode Audio data to send to the server"""
+    return base64.b64encode(data.tobytes()).decode("UTF-8")
+
+
+class GeminiHandler(AsyncStreamHandler):
+    """Handler for the Gemini API"""
+
+    def __init__(
+        self,
+        expected_layout: Literal["mono"] = "mono",
+        output_sample_rate: int = 24000,
+        output_frame_size: int = 480,
+    ) -> None:
+        super().__init__(
+            expected_layout,
+            output_sample_rate,
+            output_frame_size,
+            input_sample_rate=16000,
+        )
+        self.client: genai.Client | None = None
+        self.input_queue: asyncio.Queue = asyncio.Queue()
+        self.output_queue: asyncio.Queue = asyncio.Queue()
+        self.quit: asyncio.Event = asyncio.Event()
+
+    def copy(self) -> "GeminiHandler":
+        """Required implementation of the copy method for AsyncStreamHandler"""
+        return GeminiHandler(
+            expected_layout=self.expected_layout,
+            output_sample_rate=self.output_sample_rate,
+            output_frame_size=self.output_frame_size,
+        )
+
+    async def stream(self) -> AsyncGenerator[bytes, None]:
+        """Helper method to stream input audio to the server. Used in start_stream."""
+        while not self.quit.is_set():
+            audio = await self.input_queue.get()
+            yield audio
+        return
+
+    async def connect(
+        self, project_id: str, location: str, voice_name: str = "Puck"
+    ) -> AsyncGenerator[bytes, None]:
+        """Connect to to genai server and start the stream"""
+        client = genai.Client(vertexai=True, project=project_id, location=location)
```

Comment:
Since you made `client` a field on the `GeminiHandler` class, should you assign this to `self.client`?

Reply:
I'll remove the field since it's only used in this method. Thanks, good catch!